### ISSUE TITLE
Support Rate Limited RPCs for Statesync

### ISSUE THREAD

**Issue by chillyvee** [tendermint/tendermint#9797]
## Summary

During statesync, RPC servers may rate limit requests

it does not appear the RPC server is actually too busy, but rather manually rate limited at NGINX to reduce abuse by the public.

## Problem Definition

Polkachu has a path to minimally reproduce the problem on the evmos network

A rate limit problem appears like this:
```
INF error from light block request from primary, removing... error="error unmarshalling: invalid character '<' looking for beginning of value" height=1252684 module=light primary={}
ERR error on light block request from witness, removing... error="error unmarshalling: invalid character '<' looking for beginning of value" module=light primary={}
```

Where the actual response body is:
```
<html>
<head><title>429 Too Many Requests</title></head>
<body>
<center><h1>429 Too Many Requests</h1></center>
<hr><center>nginx/1.18.0 (Ubuntu)</center>
</body>
</html>
```

This is especially the case for high abuse public endpoints such as Evmos.

For example these POST requests all are sent within 1 second

client.Call: address: https://rpc-evmos-ia.cosmosia.notional.ventures:443, body: {"jsonrpc":"2.0","id":3,"method":"commit","params":{"height":"7730001"}}
client.Call: address: https://rpc-evmos-ia.cosmosia.notional.ventures:443, body: {"jsonrpc":"2.0","id":4,"method":"validators","params":{"height":"7730001","page":"1","per_page":"100"}}
client.Call: address: https://rpc-evmos-ia.cosmosia.notional.ventures:443, body: {"jsonrpc":"2.0","id":5,"method":"validators","params":{"height":"7730001","page":"2","per_page":"100"}}
client.Call: address: https://rpc-evmos-ia.cosmosia.notional.ventures:443, body: {"jsonrpc":"2.0","id":3,"method":"commit","params":{"height":"7730001"}}
client.Call: address: https://rpc-evmos-ia.cosmosia.notional.ventures:443, body: {"jsonrpc":"2.0","id":4,"method":"validators","params":{"height":"7730001","page":"1","per_page":"100"}}
client.Call: address: https://rpc-evmos-ia.cosmosia.notional.ventures:443, body: {"jsonrpc":"2.0","id":5,"method":"validators","params":{"height":"7730001","page":"2","per_page":"100"}}
client.Call: address: https://rpc-evmos-ia.cosmosia.notional.ventures:443, body: {"jsonrpc":"2.0","id":6,"method":"commit","params":{"height":"7730002"}}
client.Call: address: https://rpc-evmos-ia.cosmosia.notional.ventures:443, body: {"jsonrpc":"2.0","id":7,"method":"validators","params":{"height":"7730002","page":"1","per_page":"100"}}
client.Call: address: https://rpc-evmos-ia.cosmosia.notional.ventures:443, body: {"jsonrpc":"2.0","id":8,"method":"validators","params":{"height":"7730002","page":"2","per_page":"100"}}
client.Call: address: https://rpc-evmos-ia.cosmosia.notional.ventures:443, body: {"jsonrpc":"2.0","id":6,"method":"commit","params":{"height":"7730002"}}
client.Call: address: https://rpc-evmos-ia.cosmosia.notional.ventures:443, body: {"jsonrpc":"2.0","id":7,"method":"validators","params":{"height":"7730002","page":"1","per_page":"100"}}
client.Call: address: https://rpc-evmos-ia.cosmosia.notional.ventures:443, body: {"jsonrpc":"2.0","id":8,"method":"validators","params":{"height":"7730002","page":"2","per_page":"100"}}

Reproduction Instructions:
https://polkachu.com/state_sync/evmos

The RPC to use is:
https://evmos-rpc.polkachu.com:443

This RPC also has similar problems:
https://rpc-evmos-ia.cosmosia.notional.ventures:443

The Peer to use is:
425b9969981dd38bbc53c0f3f01593fdca631667@65.108.131.174:13456

Let's see if we can get Polkachu to give any additional details such as nginx configuration.

## Proposal

Determine what rate limiting mechanism (if any) will be helpful.

**Comment by thanethomson** [comment]
The core problem you're capturing here seems to actually relate to your Nginx configuration, and not to Tendermint.

What's the impact for you of not having some kind of rate limiting at all? i.e. what happens if you keep Nginx in front of your Tendermint RPC, but turn off the rate limits?

**Comment by chillyvee** [comment]
Without Nginx, all the Nginx rate limit errors responses go away.

Without Nginx, Public RPC nodes get abused and go down and are completely unavailable for certain networks.

**Comment by chillyvee** [comment]
Do you know if all statesync queries are definitively POSTS? Or are there GETs that I might have missed?  Maybe we could see if relaxing POST rate limits solves the problem?

But this could open up public API/RPCs to POST batch abuse.

**Comment by dylanschultzie** [comment]
At Lavender.Five, we run public api, rpc, and grpc endpoints for almost every tendermint-based network we're on.

Without rate limiting, the nodes are heavily, HEAVILY abused. We average more than 1M requests per day, per network, with rate limiting. It's an order of magnitude more without rate limiting, and causes the nodes to falter.

**Comment by ericjohncarlson** [comment]
I guess the question is if state-sync traffic can be identified in some way that allows a reverse proxy to allow that traffic but still allow the ability to rate limit (or block) other rpc traffic.  Is there any inspection that can be performed at the proxy.

**Comment by thanethomson** [comment]
> Do you know if all statesync queries are definitively POSTS? Or are there GETs that I might have missed? Maybe we could see if relaxing POST rate limits solves the problem?

State sync currently uses a light client-based state provider, as per https://github.com/tendermint/tendermint/blob/e941ad9294beb3fb99af9d4db887a8c1ca0d4bc2/statesync/stateprovider.go#L48

Once the light client establishes (via the RPC) the details of the state that needs to be sync'd, the state sync continues via the P2P layer.

The problem here is not the server, and the solution is definitely not to remove rate limiting. It's a good thing to have Nginx or some other reverse proxy take care of that for you.

What's necessary, however, is for the light client to handle the rate limiting response (in this case, [HTTP code 429](https://developer.mozilla.org/en-US/docs/Web/HTTP/Status/429)) in an appropriate manner. I'll speak to the rest of the team about this and see what we can do along those lines.

A possible (yet-to-be-proven) longer-term solution is the implementation of a P2P-based light client. But it's not clear to me yet whether a P2P-based light client would actually worsen the problem encountered here of nodes being overloaded by these sorts of requests.

**Comment by chillyvee** [comment]
All great ideas, each with a different difficulty and timeline.

Let us know which ones would make sense and even if the core team doesn't have time to address it immediately, someone else might contribute code in the meantime.

**Comment by PolkachuIntern** [comment]
@chillyvee The rate limit of our RPC/API endpoints is typically 30 requests per minute on a rolling basis.

It seems that this rate-limiting setting works for most chains as we state-sync our nodes periodically without issues. The problem probably occurs to chains with many chunks to download (like Evmos).

A potential solution is to handle the 429 more gracefully by falling back to the secondary state-sync RPC node (assuming the user uses two different state-sync RPC nodes in the config.toml file.

**Comment by artifactstaking** [comment]
I agree with the "non rate limited public RPCs get abused" comments posted by others. We run HAProxy in front of our RPC clusters and run 10 requests every ten seconds, per IP. We primarily run RPCs on DEXs and Bridges. Without that rate limit, it is only a matter of time before our RPC endpoints get shredded by arbitrage bots. In this unrestricted, frenzied state, the average RPC user isn't able to access our endpoints. We typically average about 1.2 million request per day, per chain.

**Comment by peterbourgon** [comment]
> I agree with the "non rate limited public RPCs get abused" comments posted by others. We run HAProxy in front of our RPC clusters and run 10 requests every ten seconds, per IP. We primarily run RPCs on DEXs and Bridges. Without that rate limit, it is only a matter of time before our RPC endpoints get shredded by arbitrage bots. In this unrestricted, frenzied state, the average RPC user isn't able to access our endpoints.

If you offer an endpoint to the public internet without authentication or validation of inbound requests in any way, you're obviously gonna get rekt. It's not the responsibility of the program servicing those requests to manage flow control or throttling or rate limiting or any of that stuff directly. It can't be. It doesn't scale. It's _your_ responsibility as an operator to manage those concerns at a layer well above the individual processes that actually serve requests. Rate limiting and all that stuff should occur well before requests hit the full nodes themselves. HAProxy is a good start to that!

> We typically average about 1.2 million request per day, per chain.

1.2M RPD = 50k RPH = 833 RPM = 13.8 RPS which is basically trivial scale. You can support that with a `bash` script in a `for` loop. Any reasonable server should be able to yield 10k RPS as a baseline.

**Comment by artifactstaking** [comment]
Not a helpful comment in relation to state synch. Rate limiting is important and does occur before the nodes are hit themselves. The issue as currently seen is two fold: (1) That state synch rapid fires RPC calls and (2) most, if not all, public RPC endpoints are rate limited. The combination of 1 and 2 can inhibit state synchs. 

Read or reread Chilly's comments above, they are asking about posts vs gets. That question is leaning towards treating traffic differently based on its content. Wouldn't it be great to be able to differentiate state synch traffic from other traffic? I know I would like that. I could better manage flow control and throttling on my non-authenticated public RPC endpoints.

**Comment by chillyvee** [comment]
> It's _your_ responsibility as an operator to manage those concerns at a layer well above the individual processes that actually serve requests.

Yes, I think everyone here agrees and has done that.  That's why we now have a problem with statesync clients unable to use the RPC servers due to throttling.

**Comment by thanethomson** [comment]
> It's your responsibility as an operator to manage those concerns at a layer well above the individual processes that actually serve requests.

@peterbourgon I agree here.

At the same time, I think the light client needs to perhaps have a slightly greater robustness to tolerating such rate limiting. But that may take some time to discuss.

In the meantime, is everybody here making use of the new `Cache-Control` headers returned by some of the RPC endpoints available in the latest Tendermint releases? (available from [v0.34.23](https://github.com/tendermint/tendermint/blob/main/CHANGELOG.md#v03423)). With appropriate caching configuration, you should be able to serve more requests to light clients.
